File: framework/backend/go/k1s0-domain-event/outbox/relay.go
package outbox

import (
	"context"
	"encoding/json"
	"log"
	"time"

	domainevent "github.com/RyuheiKiso/k1s0/framework/backend/go/k1s0-domain-event"
)

// Relay polls the outbox store and publishes pending entries.
type Relay struct {
	store        Store
	publisher    domainevent.EventPublisher
	pollInterval time.Duration
	batchSize    int
	maxRetries   int
}

// RelayOption configures a Relay.
type RelayOption func(*Relay)

// WithPollInterval sets the polling interval.
func WithPollInterval(d time.Duration) RelayOption {
	return func(r *Relay) { r.pollInterval = d }
}

// WithBatchSize sets the batch size for fetching pending entries.
func WithBatchSize(n int) RelayOption {
	return func(r *Relay) { r.batchSize = n }
}

// WithMaxRetries sets the maximum retry count before marking as failed.
func WithMaxRetries(n int) RelayOption {
	return func(r *Relay) { r.maxRetries = n }
}

// NewRelay creates a new outbox Relay with the given options.
func NewRelay(store Store, publisher domainevent.EventPublisher, opts ...RelayOption) *Relay {
	r := &Relay{
		store:        store,
		publisher:    publisher,
		pollInterval: 5 * time.Second,
		batchSize:    100,
		maxRetries:   3,
	}
	for _, opt := range opts {
		opt(r)
	}
	return r
}

// Run starts the relay loop. It blocks until the context is cancelled.
func (r *Relay) Run(ctx context.Context) {
	ticker := time.NewTicker(r.pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := r.processBatch(ctx); err != nil {
				log.Printf("outbox relay batch error: %v", err)
			}
		}
	}
}

func (r *Relay) processBatch(ctx context.Context) error {
	entries, err := r.store.FetchPending(ctx, r.batchSize)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		envelope := &domainevent.EventEnvelope{
			EventType: entry.EventType,
			Metadata:  domainevent.NewEventMetadata("outbox-relay"),
			Payload:   json.RawMessage(entry.Payload),
		}

		if pubErr := r.publisher.Publish(ctx, envelope); pubErr != nil {
			errMsg := pubErr.Error()
			if entry.RetryCount >= r.maxRetries {
				log.Printf("outbox entry %s exceeded max retries", entry.ID)
			}
			if markErr := r.store.MarkFailed(ctx, entry.ID, errMsg); markErr != nil {
				return markErr
			}
			continue
		}

		if markErr := r.store.MarkPublished(ctx, entry.ID); markErr != nil {
			return markErr
		}
	}

	return nil
}

File: framework/backend/rust/crates/k1s0-domain-event/src/outbox/relay.rs
use std::sync::Arc;
use std::time::Duration;

use tracing::{debug, error, info};

use super::store::OutboxStore;
use crate::error::OutboxError;
use crate::publisher::EventPublisher;

/// Outbox テーブルからイベントをポーリングして発行するリレー。
pub struct OutboxRelay {
    store: Arc<dyn OutboxStore>,
    publisher: Arc<dyn EventPublisher>,
    poll_interval: Duration,
    batch_size: i64,
    max_retries: i32,
}

impl OutboxRelay {
    /// ビルダーを返す。
    #[must_use]
    pub fn builder(
        store: Arc<dyn OutboxStore>,
        publisher: Arc<dyn EventPublisher>,
    ) -> OutboxRelayBuilder {
        OutboxRelayBuilder {
            store,
            publisher,
            poll_interval: Duration::from_secs(5),
            batch_size: 100,
            max_retries: 3,
        }
    }

    /// リレーを開始する。キャンセルされるまでポーリングを続ける。
    pub async fn run(&self, mut cancel: tokio::sync::oneshot::Receiver<()>) {
        info!(
            poll_interval_ms = self.poll_interval.as_millis() as u64,
            batch_size = self.batch_size,
            "outbox relay started"
        );

        loop {
            tokio::select! {
                _ = &mut cancel => {
                    info!("outbox relay stopping");
                    break;
                }
                _ = tokio::time::sleep(self.poll_interval) => {
                    if let Err(e) = self.process_batch().await {
                        error!(error = %e, "outbox relay batch failed");
                    }
                }
            }
        }
    }

    async fn process_batch(&self) -> Result<(), OutboxError> {
        let entries = self.store.fetch_pending(self.batch_size).await?;
        if entries.is_empty() {
            return Ok(());
        }

        debug!(count = entries.len(), "processing outbox entries");

        for entry in entries {
            let envelope = crate::envelope::EventEnvelope {
                event_type: entry.event_type.clone(),
                metadata: crate::envelope::EventMetadata::new("outbox-relay"),
                payload: entry.payload.clone(),
            };

            match self.publisher.publish(envelope).await {
                Ok(()) => {
                    self.store.mark_published(entry.id).await?;
                }
                Err(e) => {
                    let error_msg = e.to_string();
                    if entry.retry_count >= self.max_retries {
                        error!(
                            entry_id = %entry.id,
                            retries = entry.retry_count,
                            "outbox entry exceeded max retries, marking failed"
                        );
                        self.store.mark_failed(entry.id, &error_msg).await?;
                    } else {
                        self.store.mark_failed(entry.id, &error_msg).await?;
                    }
                }
            }
        }

        Ok(())
    }
}

/// `OutboxRelay` のビルダー。
pub struct OutboxRelayBuilder {
    store: Arc<dyn OutboxStore>,
    publisher: Arc<dyn EventPublisher>,
    poll_interval: Duration,
    batch_size: i64,
    max_retries: i32,
}

impl OutboxRelayBuilder {
    /// ポーリング間隔を設定する。
    #[must_use]
    pub fn poll_interval(mut self, interval: Duration) -> Self {
        self.poll_interval = interval;
        self
    }

    /// バッチサイズを設定する。
    #[must_use]
    pub fn batch_size(mut self, size: i64) -> Self {
        self.batch_size = size;
        self
    }

    /// 最大リトライ回数を設定する。
    #[must_use]
    pub fn max_retries(mut self, retries: i32) -> Self {
        self.max_retries = retries;
        self
    }

    /// `OutboxRelay` を構築する。
    #[must_use]
    pub fn build(self) -> OutboxRelay {
        OutboxRelay {
            store: self.store,
            publisher: self.publisher,
            poll_interval: self.poll_interval,
            batch_size: self.batch_size,
            max_retries: self.max_retries,
        }
    }
}
